sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -211,6 +215,71 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     )
   }
 
+  private def groupingSetToSQL(
+      plan: Aggregate,
+      expand: Expand,
+      project: Project): String = {
+    // In cube/rollup/groupingsets, Analyzer creates new aliases for all group by expressions.
+    // Since conversion from attribute back SQL ignore expression IDs, the alias of attribute
+    // references are ignored in aliasMap
+    val aliasMap = AttributeMap(project.projectList.collect {
+      case a @ Alias(child, name) if !child.isInstanceOf[AttributeReference] => (a.toAttribute, a)
+    })
+
+    val groupingExprs = plan.groupingExpressions.filterNot {
+      // VirtualColumn.groupingIdName is added by Analyzer, and thus remove it.
+      case a: NamedExpression => a.name == VirtualColumn.groupingIdName
+      case o => false
+    }.map {
+      case a: AttributeReference if aliasMap.contains(a) => aliasMap(a).child
+      case o => o
+    }
+
+    val groupingSQL = groupingExprs.map(_.sql).mkString(", ")
+
+    val groupingSet = expand.projections.map(_.filter {
+      case _: Literal => false
+      case e: Expression if plan.groupingExpressions.exists(_.semanticEquals(e)) => true
+      case _ => false
+    }.map {
+      case a: AttributeReference if aliasMap.contains(a) => aliasMap(a).child
+      case o => o
+    })
+
+    val aggExprs = plan.aggregateExpressions.map {
+      case a @ Alias(child: AttributeReference, name)
+          if child.name == VirtualColumn.groupingIdName =>
```

Comment:
instead of checking the name, how about we get the `gid` first, i.e. `val gid = expand.output.last`, then we can simplify this case: `case a @ Alias(child, name) if child eq gid =>`